core-bundle/contao/library/Contao/StringUtil.php

```diff
@@ -1129,4 +1129,12 @@ public static function numberToString(float|int $number, int|null $precision = n
 
 		return rtrim("$signPart$wholePart.$decimalPart", '.');
 	}
+
+	public static function resolveReferences(array $array): array
+	{
+		return array_map(
+			static fn ($value) => \is_array($value) ? self::resolveReferences($value) : $value,
+			$array,
+		);
+	}
```

Comment:
Something like this should probably work to detect self references:

```suggestion
    public static function resolveReferences(array $array): array
    {
        return self::doResolveReferences($array);
    }

    private static function doResolveReferences(array $array, array $traversed = []): array {

        return array_map(
            static function ($value) use ($array, $traversed) {
                if (in_array($array, $traversed, true)) {
                    throw new \InvalidArgumentException('Cannot dereference an array that contains a reference to itself.');
                }

                $traversed[] = $array;

                return \is_array($value) ? self::doResolveReferences($value, $traversed) : $value;
            },
            $array,
        );
    }
```

But I'm not sure if we should consider this edge case or keep it simple.

Reply:
The `in_array()` check would be pretty expensive I think. I’d probably be in favor of keeping it simple and fast ☺️